File: apiguard/internal/testgen/ir.go
package testgen

// ParsedIR matches the Rust ApiGuardIR JSON output.
type ParsedIR struct {
	Endpoints   []IREndpoint   `json:"endpoints"`
	AuthSchemes []IRAuthScheme `json:"auth_schemes"`
	Metadata    IRMetadata     `json:"metadata"`
}

// IREndpoint represents a single API endpoint from the Rust IR.
type IREndpoint struct {
	Path        string                 `json:"path"`
	Method      string                 `json:"method"`
	Parameters  []IRParameter          `json:"parameters"`
	RequestBody *IRRequestBody         `json:"request_body,omitempty"`
	Responses   map[string]IRResponse  `json:"responses"`
	Security    []string               `json:"security"`
	Tags        []string               `json:"tags"`
	XApiguard   map[string]interface{} `json:"x_apiguard"`
}

// IRParameter represents a parameter from the Rust IR.
type IRParameter struct {
	Name     string  `json:"name"`
	Location string  `json:"location"`
	Required bool    `json:"required"`
	Schema   *string `json:"schema,omitempty"`
}

// IRRequestBody represents a request body from the Rust IR.
type IRRequestBody struct {
	ContentTypes []string `json:"content_types"`
	Schema       *string  `json:"schema,omitempty"`
	Required     bool     `json:"required"`
}

// IRResponse represents a response from the Rust IR.
type IRResponse struct {
	Description string  `json:"description"`
	Schema      *string `json:"schema,omitempty"`
}

// IRAuthScheme represents an authentication scheme from the Rust IR.
type IRAuthScheme struct {
	Name       string  `json:"name"`
	SchemeType string  `json:"scheme_type"`
	Location   *string `json:"location,omitempty"`
	HeaderName *string `json:"header_name,omitempty"`
}

// IRMetadata holds API metadata from the Rust IR.
type IRMetadata struct {
	BaseURL    string `json:"base_url"`
	APIVersion string `json:"api_version"`
	SchemaHash string `json:"schema_hash"`
}

File: apiguard/rust/parser/src/ir.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// APIGuard Internal Representation.
/// This is the normalised form all downstream layers consume.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiGuardIR {
    pub endpoints: Vec<Endpoint>,
    pub auth_schemes: Vec<AuthScheme>,
    pub metadata: Metadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Endpoint {
    pub path: String,
    pub method: HttpMethod,
    pub parameters: Vec<Parameter>,
    pub request_body: Option<RequestBody>,
    pub responses: HashMap<String, Response>,
    pub security: Vec<String>,
    pub tags: Vec<String>,
    pub x_apiguard: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Parameter {
    pub name: String,
    pub location: ParameterLocation,
    pub required: bool,
    pub schema: Option<SchemaType>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestBody {
    pub content_types: Vec<String>,
    pub schema: Option<SchemaType>,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Response {
    pub description: String,
    pub schema: Option<SchemaType>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthScheme {
    pub name: String,
    pub scheme_type: AuthSchemeType,
    /// For apikey: header, query, or cookie.
    pub location: Option<String>,
    /// For apikey/bearer: the header name.
    pub header_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Metadata {
    pub base_url: String,
    pub api_version: String,
    /// SHA-256 hex digest of the raw input bytes.
    pub schema_hash: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum HttpMethod {
    Get,
    Post,
    Put,
    Patch,
    Delete,
    Head,
    Options,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ParameterLocation {
    Path,
    Query,
    Header,
    Cookie,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AuthSchemeType {
    Bearer,
    Jwt,
    OAuth2,
    ApiKey,
    Basic,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SchemaType {
    String,
    Integer,
    Number,
    Boolean,
    Array(Box<SchemaType>),
    Object(HashMap<String, SchemaType>),
    Ref(String),
}

/// Input format hint for the parser.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InputFormat {
    OpenApi3,
    Swagger2,
    GraphQL,
    Postman,
}
